docs/source/daml/intro/3_Data.rst

```diff
@@ -0,0 +1,202 @@
+
+It's a good idea to always derive ``Eq`` and ``Show`` using ``deriving (Eq, Show)``. The record types created using ``template T with`` do this automatically.
+
+Records can give the data on ``CashBalance`` a bit more structure:
+
+.. literalinclude:: daml/3_Data/Record.daml
+  :language: daml
+  :start-after: -- CASH_BALANCE_BEGIN
+  :end-before: -- CASH_BALANCE_END
+
+If you look at the resulting scenario view, you'll see that this still gives rise to one table. The records are expanded out into columns using dot notation.
+
+Variants
+~~~~~~~~
+
+Suppose now that you also wanted to keep track of cash in hand. Cash in hand doesn't have a bank, but you can't just leave ``bank`` empty. DAML doesn't have an equivalent to ``null``. Variants can express that cash can either be in hand or at a bank.
+
+.. literalinclude:: daml/3_Data/Variants.daml
+  :language: daml
+  :start-after: -- CASH_BALANCE_BEGIN
+  :end-before: -- CASH_BALANCE_END
+
+The way to read the declaration of ``Location`` is "*A Location either has value* ``InHand`` *OR has a value* ``InAccount a`` *where* ``a`` *is of type Account*". This is quite an explicit way to say that there may or may not be an ``Account`` associated with a ``CashBalance`` and gives both cases suggestive names. The generic way would be to use the built-in ``Optional`` type. The ``None`` value of type ``Optional a`` is the closest DAML has to a ``null`` value.
+
+.. literalinclude:: daml/3_Data/Variants.daml
+  :language: daml
+  :start-after: -- OPTIONAL_BEGIN
+  :end-before: -- OPTIONAL_END
+
+Variant types where none of the data constructors take a parameter are called enums:
+
+.. literalinclude:: daml/3_Data/Variants.daml
+  :language: daml
+  :start-after: -- ENUM_BEGIN
+  :end-before: -- ENUM_END
+
+Accessing data in variants requires distinguishing the different possible cases. It is no longer possible to access the account number of a ``Location`` directly, for example, as there may be no account number in case of ``InHand``. Instead, use *pattern matching* and either throw errors or return compatible types for all cases.
+
+.. literalinclude:: daml/3_Data/Variants.daml
+  :language: daml
+  :start-after: -- VARIANT_ACCESS_BEGIN
+  :end-before: -- VARIANT_ACCESS_END
+
+Manipulating Data
+-----------------
+
+You've got all the ingredients to build rich types expressing the data you want to be able to write to the ledger, and you have seen how to create new values and read fields from values. But how do you manipulate values once created?
+
+All data in DAML is immutable, meaning once a value is created, it will never change. Rather than changing values, one creates new values based on old ones with some changes applied. Whilst contracts and records may feel a bit like objects in object oriented langauges, they behave quite different in this respect.
```

Comment:
I'm not sure I would keep that second sentence; many OO language communities have some notion of immutable objects, with varying levels of technical support (`Object.freeze` in JS, compiler-detected immutable objects being thread-safe in Java, etc.). I'm not sure what the analogy adds here. If you do want to keep it, though:
```suggestion
All data in DAML is immutable, meaning once a value is created, it will never change. Rather than changing values, one creates new values based on old ones with some changes applied. Whilst contracts and records may feel a bit like objects in object-oriented languages, they behave quite different in this respect.
```